src/librustc_mir/interpret/intrinsics.rs

```diff
@@ -220,7 +218,18 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
             sym::discriminant_value => {
                 let place = self.deref_operand(args[0])?;
                 let discr_val = self.read_discriminant(place.into())?.0;
-                self.write_scalar(Scalar::from_u64(u64::try_from(discr_val).unwrap()), dest)?;
+                let scalar = match dest.layout.ty.kind {
+                    ty::Int(_) => Scalar::from_int(discr_val as i128, dest.layout.size),
```

Comment:
The `as` cast here is very suspicious. Does this really handle truncation / sign extension properly?

Reply:
@eddyb maybe `read_discriminant` should return something more typed, such as `Scalar` or even `ImmTy`?

Reply:
Afaict `u128` to `i128` is a simple bitcast, so it should be fine. Will add a test though

Reply:
The cast is a simple bitcast, but this could mean that we did the wrong sign extension or so earlier. We had critical soundness bugs in Miri where the only "tell" was such a cast. It's really easy to screw up generic handling of arbitrarily-sized and -signed integers, and I'd feel better if this code here wouldn't have to go quite so deep in our stack of abstractions.

At the very least, if you need to cast here this indicates to me that the return type of `read_discriminant` is not picked well.

Reply:
`read_discriminant` should at least return `Scalar`, I agree. `ImmTy` could also work, not sure if it's worth doing.

Reply:
@eddyb so, something like this?

https://github.com/RalfJung/rust/commit/de32d1e058a3c2801d783bd1ddebc0c2da9067d0

Some things were a bit strange. I hope what I say in the comments about there being three layouts make sense, but I had no idea what to even call them. I left a FIXME comment for something I did not understand. I also had lots of trouble with terminology between "tags" and "discriminants".